drivers/staging/vc04_services/bcm2835-codec/bcm2835-v4l2-codec.c

```diff
@@ -1260,15 +1260,37 @@ static int vidioc_g_selection(struct file *file, void *priv,
 {
 	struct bcm2835_codec_ctx *ctx = file2ctx(file);
 	struct bcm2835_codec_q_data *q_data;
-	bool capture_queue = s->type == V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE ?
-								true : false;
+	enum v4l2_buf_type type;
 
-	if ((ctx->dev->role == DECODE && !capture_queue) ||
-	    (ctx->dev->role == ENCODE && capture_queue))
-		/* OUTPUT on decoder and CAPTURE on encoder are not valid. */
+	/*
+	 * The selection API takes V4L2_BUF_TYPE_VIDEO_CAPTURE and
+	 * V4L2_BUF_TYPE_VIDEO_OUTPUT, even if the device implements the MPLANE
+	 * API. The V4L2 core will have converted the MPLANE variants to
+	 * non-MPLANE.
+	 * Open code this instead of using get_q_data in this case.
+	 */
+	switch (s->type) {
+	case V4L2_BUF_TYPE_VIDEO_CAPTURE:
+		/* CAPTURE on encoder is not valid. */
+		if (ctx->dev->role == ENCODE)
+			return -EINVAL;
+		type = V4L2_BUF_TYPE_VIDEO_CAPTURE_MPLANE;
+		q_data = &ctx->q_data[V4L2_M2M_DST];
+		break;
+	case V4L2_BUF_TYPE_VIDEO_OUTPUT:
+		/* OUTPUT on deoder is not valid. */
+		if (ctx->dev->role == DECODE)
+			return -EINVAL;
+		type = V4L2_BUF_TYPE_VIDEO_OUTPUT_MPLANE;
+		q_data = &ctx->q_data[V4L2_M2M_SRC];
+		break;
+	default:
 		return -EINVAL;
+	}
 
-	q_data = get_q_data(ctx, s->type);
+	v4l2_err(&ctx->dev->v4l2_dev, "%s: queue %u\n", __func__, type);
+
+	q_data = get_q_data(ctx, type);
```

Comment:
q_data is assigned above, and again here without any intervening use, which looks wrong. By analogy with vidioc_s_selection I think this line should be deleted.

Reply:
Quite correct, and made me check the 
```
if (!q_data)
		return -EINVAL;
```
below,. Seeing as q_data is assigned the address of an array within ctx, it can never be NULL, so no need to check it. Removed that too.

Reply:
Not to mention the debug line left behind, and ```type``` being unused.